File: cypress/e2e/generate_drm_tests.js
const fs = require('fs');
const path = require('path');

// drm_combinations.txt 파일 경로
const combinationsPath = path.join(__dirname, 'drm_combinations.txt');

// 파일 내용 읽기
const fileContent = fs.readFileSync(combinationsPath, 'utf8');
const lines = fileContent.split('\n');

// 작업타입 파싱
let currentType = 'DRM';
let currentInputCodec = '';
let currentOutputCodec = '';
let currentResolution = '';
let currentStreamingFormat = '';
let currentAspectRatio = false;
let currentAudioEncryption = false;
let currentMultiKey = false;
let currentMultiManifest = false;
let currentDrmOff = false;

const testCases = [];

// Helper function to add a test case and reset variables
function addTestCaseAndReset() {
  if (currentInputCodec && currentOutputCodec && currentResolution && currentStreamingFormat) {
    testCases.push({
      taskName: `${testCases.length + 1}_${currentType.toLowerCase()}_${currentInputCodec}_${currentOutputCodec}_${currentResolution}_${currentStreamingFormat}`,
      type: currentType,
      cid: 'test',
      inputCodec: currentInputCodec,
      outputCodec: currentOutputCodec,
      resolutionProfile: currentResolution,
      streamingFormat: currentStreamingFormat,
      aspectRatio: currentAspectRatio,
      audioEncryption: currentAudioEncryption,
      multiKey: currentMultiKey,
      multiManifest: currentMultiManifest,
      drmOff: currentDrmOff
    });
  }
  // Reset variables for the next test case
  currentInputCodec = '';
  currentOutputCodec = '';
  currentResolution = '';
  currentStreamingFormat = '';
  currentAspectRatio = false;
  currentAudioEncryption = false;
  currentMultiKey = false;
  currentMultiManifest = false;
  currentDrmOff = false;
}

for (const line of lines) {
  if (line.includes('작업타입:')) {
    // Before processing the new type, add the previously collected test case
    addTestCaseAndReset();

    // Parse the new type
    const typeMatch = line.match(/작업타입:\s*(\w+)(?:\+(\w+))?/);
    if (typeMatch) {
      if (typeMatch[2]) {
        currentType = `${typeMatch[1]}_${typeMatch[2]}`.toUpperCase();
      } else {
        currentType = typeMatch[1].toUpperCase();
      }
    }
  } else if (line.includes('입력파일:')) {
    currentInputCodec = line.split(':')[1].trim().toLowerCase();
  } else if (line.includes('출력 코덱:')) {
    currentOutputCodec = line.split(':')[1].trim().toLowerCase();
  } else if (line.includes('해상도:')) {
    currentResolution = line.split(':')[1].trim().toLowerCase().replace(/\+/g, '_');
  } else if (line.includes('스트리밍 포멧:')) {
    currentStreamingFormat = line.split(':')[1].trim().toLowerCase().replace(/\+/g, '_');
  } else if (line.includes('비율옵션:')) {
    const aspectRatioValue = line.split(':')[1].trim();
    if (aspectRatioValue === '체크') {
      currentAspectRatio = true;
    } else if (aspectRatioValue === '체크해제') {
      currentAspectRatio = false;
    }
  } else if (line.includes('오디오 암호화:')) {
    currentAudioEncryption = line.split(':')[1].trim() === '체크';
  } else if (line.includes('멀티키:')) {
    currentMultiKey = line.split(':')[1].trim() === '체크';
  } else if (line.includes('멀티매니페스트:')) {
    currentMultiManifest = line.split(':')[1].trim() === '체크';
  } else if (line.includes('DRM 비활성화 구간:')) {
    currentDrmOff = line.split(':')[1].trim() === '설정';
    // No need to push test case here anymore
  }
}

// Add the very last test case after the loop finishes
addTestCaseAndReset();

if (testCases.length === 0) {
  console.error('테스트 케이스가 생성되지 않았습니다.');
  process.exit(1);
}

// 테스트 파일 내용 생성
const testFileContent = `
// 자동 생성된 테스트 파일
before(() => {
  // 새로운 로그 파일 생성
  cy.task('generateLogFileName').then((fileName) => {
    Cypress.env('currentLogFile', fileName);
  });
//테스트 시작 로그 기록
cy.writelog('Test Start');
cy.sendTestStartMessage();
});

after(() => {
  // 테스트 종료 후 로그 마무리
  cy.writelog('Test End');
  cy.sendTestCompletionMessage(testStats);
});

${testCases.map(testCase => `
describe('${testCase.taskName}', () => {
  it('${testCase.taskName} 작업생성', function() {
    cy.viewport(1280, 720);
    cy.login('jcg@inka.co.kr', 'Say1013!');

    cy.createDRMTask({
      taskName: '${testCase.taskName}',
      cid: '${testCase.cid}',
      inputCodec: '${testCase.inputCodec.toUpperCase()}',
      resolutionProfile: '${testCase.resolutionProfile.toUpperCase()}',
      streamingFormat: '${testCase.streamingFormat.toUpperCase()}',
      outputCodec: '${testCase.outputCodec.toUpperCase()}',
      type: '${testCase.type}',
      aspectRatio: ${testCase.aspectRatio},
      audioEncryption: ${testCase.audioEncryption},
      multiKey: ${testCase.multiKey},
      multiManifest: ${testCase.multiManifest},
      drmOff: ${testCase.drmOff}
    });
  });

  afterEach(function() {
    if (this.currentTest.state === 'failed') {
      cy.writelog(\`테스트 실패: \${this.currentTest.title}\`);
      cy.screenshot(\`실패_\${this.currentTest.title}\`);
      cy.sendTestFailureMessage(this.currentTest.title, this.currentTest.err);
    } else if (this.currentTest.state === 'passed') {
      testStats.passed++;
    }
    testStats.total++;
  });
});
`).join('\n')}
`;

// 테스트 파일 저장
const outputPath = path.join(__dirname, 'drm_auto_generated_tests.cy.js');
fs.writeFileSync(outputPath, testFileContent);

console.log('테스트 파일이 생성되었습니다:', outputPath);
console.log('생성된 테스트 케이스 수:', testCases.length); 

File: cypress/e2e/generate_drm_tests.py
import os
import re
from pathlib import Path

# drm_combinations.txt 파일 경로
current_dir = Path(__file__).parent
combinations_path = current_dir / 'drm_combinations.txt'

# 파일 내용 읽기
with open(combinations_path, 'r', encoding='utf-8') as file:
    lines = file.readlines()

# 작업타입 파싱
current_type = 'DRM'
current_input_codec = ''
current_output_codec = ''
current_resolution = ''
current_streaming_format = ''
current_aspect_ratio = False
current_audio_encryption = False
current_multi_key = False
current_multi_manifest = False
current_drm_off = False

test_cases = []

def add_test_case_and_reset():
    global current_input_codec, current_output_codec, current_resolution, current_streaming_format
    global current_aspect_ratio, current_audio_encryption, current_multi_key, current_multi_manifest, current_drm_off
    
    if current_input_codec and current_output_codec and current_resolution and current_streaming_format:
        test_cases.append({
            'taskName': f"{len(test_cases) + 1}_{current_type.lower()}_{current_input_codec}_{current_output_codec}_{current_resolution}_{current_streaming_format}",
            'type': current_type,
            'cid': 'test',
            'inputCodec': current_input_codec,
            'outputCodec': current_output_codec,
            'resolutionProfile': current_resolution,
            'streamingFormat': current_streaming_format,
            'aspectRatio': current_aspect_ratio,
            'audioEncryption': current_audio_encryption,
            'multiKey': current_multi_key,
            'multiManifest': current_multi_manifest,
            'drmOff': current_drm_off
        })
    
    # Reset variables for the next test case
    current_input_codec = ''
    current_output_codec = ''
    current_resolution = ''
    current_streaming_format = ''
    current_aspect_ratio = False
    current_audio_encryption = False
    current_multi_key = False
    current_multi_manifest = False
    current_drm_off = False

for line in lines:
    line = line.strip()
    if '작업타입:' in line:
        # Before processing the new type, add the previously collected test case
        add_test_case_and_reset()

        # Parse the new type
        type_match = re.match(r'작업타입:\s*(\w+)(?:\+(\w+))?', line)
        if type_match:
            if type_match.group(2):
                current_type = f"{type_match.group(1)}_{type_match.group(2)}".upper()
            else:
                current_type = type_match.group(1).upper()
    elif '입력파일:' in line:
        current_input_codec = line.split(':')[1].strip().lower()
    elif '출력 코덱:' in line:
        current_output_codec = line.split(':')[1].strip().lower()
    elif '해상도:' in line:
        current_resolution = line.split(':')[1].strip().lower().replace('+', '_')
    elif '스트리밍 포멧:' in line:
        current_streaming_format = line.split(':')[1].strip().lower().replace('+', '_')
    elif '비율옵션:' in line:
        aspect_ratio_value = line.split(':')[1].strip()
        current_aspect_ratio = aspect_ratio_value == '체크'
    elif '오디오 암호화:' in line:
        current_audio_encryption = line.split(':')[1].strip() == '체크'
    elif '멀티키:' in line:
        current_multi_key = line.split(':')[1].strip() == '체크'
    elif '멀티매니페스트:' in line:
        current_multi_manifest = line.split(':')[1].strip() == '체크'
    elif 'DRM 비활성화 구간:' in line:
        current_drm_off = line.split(':')[1].strip() == '설정'

# Add the very last test case after the loop finishes
add_test_case_and_reset()

if not test_cases:
    print('테스트 케이스가 생성되지 않았습니다.')
    exit(1)

# 테스트 파일 내용 생성
test_file_content = f"""
const TEST_COUNT = {len(test_cases)}; // 총 테스트 수
let testStats = {{
  total: 0,
  passed: 0,
  failed: 0
}};  

// 자동 생성된 테스트 파일
before(() => {{
  // 새로운 로그 파일 생성
  cy.task('generateLogFileName').then((fileName) => {{
    Cypress.env('currentLogFile', fileName);
  }});
//테스트 시작 로그 기록
cy.writelog('Test Start');
cy.sendTestStartMessage('🚀 DRM 자동화 테스트가 시작되었습니다.');
}});

beforeEach(function() {{
  cy.writelog(`생성 시작: ${{this.currentTest.title}}`);
}});

after(() => {{
  // 테스트 종료 후 로그 마무리
  cy.writelog('Test End');
  cy.sendTestCompletionMessage(testStats);
}});

"""

for test_case in test_cases:
    test_file_content += f"""
describe('{test_case['taskName']}', () => {{
  it('{test_case['taskName']} 작업생성', function() {{
    cy.viewport(1280, 720);
    cy.login('jcg@inka.co.kr', 'Say1013!');

    cy.createDRMTask({{
      taskName: '{test_case['taskName']}',
      cid: '{test_case['cid']}',
      inputCodec: '{test_case['inputCodec'].upper()}',
      resolutionProfile: '{test_case['resolutionProfile'].upper()}',
      streamingFormat: '{test_case['streamingFormat'].upper()}',
      outputCodec: '{test_case['outputCodec'].upper()}',
      type: '{test_case['type']}',
      aspectRatio: {str(test_case['aspectRatio']).lower()},
      audioEncryption: {str(test_case['audioEncryption']).lower()},
      multiKey: {str(test_case['multiKey']).lower()},
      multiManifest: {str(test_case['multiManifest']).lower()},
      drmOff: {str(test_case['drmOff']).lower()}
    }});
  }});

  afterEach(function() {{
    if (this.currentTest.state === 'failed') {{
      cy.writelog(`테스트 실패: ${{this.currentTest.title}}`);
      cy.screenshot(`실패_${{this.currentTest.title}}`);
      cy.sendTestFailureMessage(this.currentTest.title, this.currentTest.err);
    }} else if (this.currentTest.state === 'passed') {{
      testStats.passed++;
    }}
    testStats.total++;
  }});
}});
"""

# 테스트 파일 저장
output_path = current_dir / 'drm_auto_generated_tests.cy.js'
with open(output_path, 'w', encoding='utf-8') as file:
    file.write(test_file_content)

print('테스트 파일이 생성되었습니다:', output_path)
print('생성된 테스트 케이스 수:', len(test_cases)) 